### PR TITLE
Implement Display for Objects

### PR DESCRIPTION
**Pull request by IovoslavIovchev**
**NB:** Due to the fact that the main binary uses the `dbg!` macro, pretty printing, while present, does not the make output, well.. _pretty_ (e.g. new lines are rendered as `\n`).

### CODE REVIEW

**Comment by IovoslavIovchev** [comment]
#170

**Comment by jasonwilliams** [comment]
This may be out of scope for your PR, but be weary of circular references.
Printing an object A which has a property B which has a property back to A I think causes an infinite loop.

**Comment by IovoslavIovchev** [comment]
This implementation indeed does not take circular dependencies into consideration. 

I believe I have an idea how that can be achieved, so I will take some more time to look into it and deliver everything in one PR.

**Comment by IovoslavIovchev** [comment]
@jasonwilliams could you take a look?

**Comment by IovoslavIovchev** [comment]
I decided to not filter out the prototype slot, as I still have not found a scenario which causes a stack overflow.
I propose we leave it as is, and if needed, remove it later.

**Comment by jasonwilliams** [comment]
Do we want to somehow share this logic with `console.log` https://github.com/jasonwilliams/boa/blob/master/src/lib/js/console.rs ?

**Comment by IovoslavIovchev** [comment]
We definitely could do that, only we need to find a way to escape code duplication for printing properties.

**Comment by jasonwilliams** [comment]
Not sure if you're aware but the newlines aren't being parsed
This is the output i get
```
[src/bin/bin.rs:57] exec(&buffer) = "{\n\n    __proto__: {\n        valueOf: function() { [native code] },\n        constructor: {\n            prototype: \'[Cycle]\'\n            call: function() { [native code] },\n            construct: function() { [native code] }\n        },\n        toString: function() { [native code] }\n        BooleanData: false,\n        extensible: true,\n        __proto__: undefined\n    },\n    BooleanData: false\n}"
```

**Comment by IovoslavIovchev** [comment]
@jasonwilliams As I have mentioned in the PR description, this is because of the `dbg!` macro. It does not render new lines or tabs, etc. but rather `\n`, `\t`, etc.

If you used `println!` rather than `dgb!` on ln 57 in `src/bin/bin.rs`, you would see the output beautified. Hope that helps.

**Comment by jasonwilliams** [comment]
@IovoslavIovchev oh my bad.
Should we change bin.rs to print() instead of dbg! then?

**Comment by IovoslavIovchev** [comment]
Well, that really depends. In my opinion we should not, as bin.rs is used for debugging purposes anyway, and it is better to see that the output is what you actually expect.

The downside obviously is that when it wraps (the way it does in your case) it gets really hard to read.

**Comment by IovoslavIovchev** [comment]
Additionally, one slight mistake I made was I joined the properties and slots before printing and that resulted in
```
        valueOf: function() { [native code] },     
        toString: function() { [native code] }     
        BooleanData: false,
```
as you can see there is no comma after `toString`.

However, I have fixed it now. You can review the latest change.

**Comment by jasonwilliams** [comment]
Sure, you also have a conflict

**Comment by IovoslavIovchev** [comment]
@jasonwilliams Regarding sharing some functionality with `console.log`, the way I see it, we could either create a macro (I am hoping this way we would be able to eliminate _even_ the code duplication for slots and properties) or have two separate functions that would return `Vec<String>`s, one of which we would be able to reuse for `console.log`.

What are your thoughts on that? I am open to any suggestions.

**Comment by jasonwilliams** [comment]
I’m liking the sound of the macro idea

**Comment by IovoslavIovchev** [comment]
I will take a look at what can be done

**Comment by jasonwilliams** [comment]
@IovoslavIovchev here is an example of circular deps printing forever
https://gist.github.com/jasonwilliams/62616cf8a1529411f681e67aec847fa0

That's the global object, its all fine until it hits `String` which has a prototype, and the prototype has a `constructor` property which points back to `String`, then `String`s "prototype property points back to the prototype object, and it goes on infinitely

**Comment by jasonwilliams** [comment]
@IovoslavIovchev is this PR still WIP?

**Comment by IovoslavIovchev** [comment]
Yes, it is. Unfortunately, I haven't found an adequate way to do what I want, and am still researching.

**Comment by jasonwilliams** [comment]
No worries, was just curious

**Comment by jasonwilliams** [comment]
You need any help with this?

**Comment by IovoslavIovchev** [comment]
I hit a wall with `prototype`s. When the execution reaches `length` (for instance when printing `Array.prototype`) I get a `None` and the program panics.

File: src/lib/builtins/array.rs [hunk 1]
```diff
@@ -698,25 +700,25 @@ mod tests {
...
-        let _ne = forward(&mut engine, "one.concat(empty)");
-        //assert_eq!(ne, String::from("a.b.c"));
-        // NonEmpty ++ NonEmpty
-        let _nn = forward(&mut engine, "one.concat(one)");
-        //assert_eq!(nn, String::from("a.b.c"));
+        // let realm = Realm::create();
+        // let mut engine = Executor::new(realm);
+        // let init = r#"
+        // var empty = new Array();
+        // var one = new Array(1);
```
Comment (IovoslavIovchev) [inline src/lib/builtins/array.rs]: Currently, `new Array(1)` produces `[1]` (an array with one element, which is `1`), when, in fact, it should produce an array with `length` 1 and its first element should be `undefined`.

**Comment by IovoslavIovchev** [comment]
@jasonwilliams could you take a look again?

File: src/lib/builtins/array.rs [hunk 2]
```diff
@@ -326,14 +326,14 @@
~            "missing callback when calling function Array.prototype.every".to_string(),
~        ));
~    }
~    let callback = &args[0];
~    let this_arg = if args.len() > 1 {
~        args[1].clone()
~    } else {
```
Comment (jasonwilliams) [inline src/lib/builtins/array.rs]: Looks like `Array.prototype.length` shouldn't have a getter according to the spec

File: src/lib/builtins/array.rs (hunk 1 above)
Comment (jasonwilliams) [reply src/lib/builtins/array.rs]: We will pick this up in https://github.com/jasonwilliams/boa/issues/209

File: src/lib/builtins/string.rs [hunk 3]
```diff
@@ -741,7 +741,9 @@ pub fn create_constructor(global: &Value) -> Value {
 
     // Create prototype
     let proto = ValueData::new_obj(Some(global));
-    let prop = Property::default().get(to_value(get_string_length as NativeFunctionData));
+    let prop = Property::default()
+        .get(to_value(get_string_length as NativeFunctionData))
```
Comment (jasonwilliams) [inline src/lib/builtins/string.rs]: `String.prototype.length` does not have a getter according to the spec https://tc39.es/ecma262/#sec-properties-of-the-string-prototype-object

**Pull request closed without merging**